File: LeetCode/困难/239. 滑动窗口最大值/Solution239.java
/*
url: https://leetcode.cn/problems/sliding-window-maximum/
LeetCode参考: https://leetcode.cn/problems/sliding-window-maximum/solutions/543426/hua-dong-chuang-kou-zui-da-zhi-by-leetco-ki6m/
标签: 单调队列, 用单调队列求区间内最大值
*/

import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;
import java.util.PriorityQueue;

public class Solution239 {
    // 使用单调队列
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int[] ans = new int[n - k + 1];
        Deque<Integer> q = new LinkedList<>(); // 单调(往队首递增)队列，用于求解区间内的最大值
        // 初始的队列
        for (int i = 0; i < k; i++) {
            // 新元素i入队，**注意**，由于i是目前最靠右的，也就是说是最后移除的，以下移除的元素<=nums[i]且比i更快出队，当不了最大值，因此可以放心移除
            while (!q.isEmpty() && nums[q.peekLast()] <= nums[i]) {
                q.pollLast();
            }
            q.offerLast(i);
        }
        ans[0] = nums[q.peek()];
        for (int i = k; i < n; i++) {
            while (!q.isEmpty() && nums[q.peekLast()] <= nums[i]) {
                q.pollLast();
            }
            q.offerLast(i);
            // 此时区间为[i - k + 1, i]，求最大值时，若不在区间范围内则全部移除
            while (q.peekFirst() < i - k + 1) {
                q.pollFirst();
            }
            ans[i - k + 1] = nums[q.peek()];
        }
        return ans;
    }

    // 使用优先队列的解法，不比上面用单调队列的更优，但是也实现一下。使用优先队列，若nums元素单增，不会删除元素，查询最大元素共需要log1+log2+...+logn=lonn!=O(nlogn)时间
    public int[] maxSlidingWindow_implementation2(int[] nums, int k) {
        int n = nums.length;
        PriorityQueue<Integer> q = new PriorityQueue<>((x, y) -> {
            return nums[y] - nums[x];
        });
        int[] ans = new int[n - k + 1];
        for (int i = 0; i < n; i++) { // i为窗口右端点
            if (i < k - 1) {
                q.offer(i);
            }
            else {
                q.offer(i);
                while (q.peek() < i - k + 1) {
                    q.poll();
                }
                ans[i - k + 1] = nums[q.peek()];
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        Solution239 solu = new Solution239();
        System.out.println(Arrays.toString(solu.maxSlidingWindow(new int[]{7,2,4}, 2)));
        System.out.println(Arrays.toString(solu.maxSlidingWindow_implementation2(new int[]{7,2,4}, 2)));
    }
}

File: LeetCode/困难/239. 滑动窗口最大值/solution239.cpp
/*
url: https://leetcode.cn/problems/sliding-window-maximum/
相关: LeetCode1438. 绝对差不超过限制的最长连续子数组, LeetCode3835. 开销小于等于 K 的子数组数目
*/

#include <bits/stdc++.h>

using namespace std;

class Solution239 {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> ans(n - k + 1);
        deque<int> dq; // 双端队列，从左往右单减
        for (int l = 0, r = 0; r < n; r++) {
            /* 入尾，注意r必须从尾部往前排除比自己小的元素，而不是从头部往后排除。例如，deque现在数字是[1 5 2 2]，要进来一个3，把1排出
            后，后面窗口滑动，5出去，到了[2 2 3]时，会认为2是当前最大值，而实际应当是3。
            */
            while (!dq.empty() && nums[dq.back()] <= nums[r]) dq.pop_back();
            dq.push_back(r);
            if (r >= k - 1) {
                // 出头，获取最大值时先排出头部过时元素
                while (dq.front() < l) dq.pop_front();
                ans[l] = nums[dq.front()];
                l++;
            }
        }
        return ans;
    }
};

int main(int argc, char const *argv[]) {
    Solution239 solu;
    // auto a = solu.maxSlidingWindow(*new vector<int>{1,3,-1,-3,5,3,6,7}, 3);
    auto b = solu.maxSlidingWindow(*new vector<int>{1,3,1,2,0,5}, 3);
    return 0;
}